ipns/name.go

```diff
@@ -0,0 +1,94 @@
+package ipns
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/multiformats/go-multibase"
+)
+
+// Name represents an [IPNS Name].
+//
+// [IPNS Name]: https://specs.ipfs.tech/ipns/ipns-record/#ipns-name
+type Name struct {
+	src []byte
+}
+
+// NameFromString creates a [Name] from the given IPNS Name in its [string representation].
+//
+// [string representation]: https://specs.ipfs.tech/ipns/ipns-record/#string-representation
+func NameFromString(str string) (Name, error) {
+	str = strings.TrimPrefix(str, "/ipns/")
+	pid, err := peer.Decode(str)
+	if err != nil {
+		return Name{}, err
+	}
+	return NameFromPeer(pid), nil
+}
+
+// NameFromRoutingKey creates a [Name] from the given IPNS Name in its routing key
+// representation. See [Name.RoutingKey] for more information.
+func NameFromRoutingKey(data []byte) (Name, error) {
+	if !bytes.HasPrefix(data, []byte("/ipns/")) {
+		return Name{}, ErrInvalidName
+	}
+
+	data = bytes.TrimPrefix(data, []byte("/ipns/"))
+	pid, err := peer.IDFromBytes(data)
+	if err != nil {
+		return Name{}, err
+	}
+	return NameFromPeer(pid), nil
+}
+
+// NameFromPeer creates a [Name] from the given [peer.ID].
+func NameFromPeer(pid peer.ID) Name {
+	return Name{src: []byte(pid)}
+}
+
+// RoutingKey returns the binary IPNS Routing Key for the given [Name]. Note that
+// the intended use of this function is for routing purposes only. The output of
+// this function is binary, not human readable. For a human-readable string, see [Name.Key].
+func (n Name) RoutingKey() []byte {
+	var buffer bytes.Buffer
+	buffer.WriteString("/ipns/")
+	buffer.Write(n.src)
+	return buffer.Bytes()
+}
+
+// String returns the human-readable IPNS Name, prefixed with "/ipns/".
+func (n Name) String() string {
+	cid := peer.ToCid(n.pid())
```

Comment:
:thought_balloon: thoughts on removing dep. on `peer.ToCid` and adding ipns-specific func?

```go
// Cid encodes IPNS Name as a CID of the public key.
//
// If the IPNS Name is invalid (e.g., empty), this will return the empty CID.
func (n Name) Cid() cid.Cid {
       m, err := mh.Cast([]byte(n))
       if err != nil {
               return cid.Undef
       }
       return cid.NewCidV1(cid.Libp2pKey, m)
}
```

while at it we could also:

```go
// FromCid converts a CID to an IPNS Name, if possible.
func NameFromCid(c cid.Cid) (Name, error) {
       code := mc.Code(c.Type())
       if code != mc.Libp2pKey {
               return "", fmt.Errorf("CID codec %q is not allowed for IPNS Names, use  %q instead", code, mc.Libp2pKey)
       }
       return Name(c.Hash()), nil
}
```

I feel we will need them anyway, see places like https://github.com/ipfs/boxo/pull/333#pullrequestreview-1480332571